### PR TITLE
Cached sequencer view implementation

### PR DESCRIPTION
**Pull request by no2chem**
## Overview

Description: This PR introduces a cached version of the sequencer view, which enables the sequencer view to remember tails it has learned about. This reduces calls to the sequencer at a small cost of local client memory (roughly equal to 32 bytes per stream).

Why should this be merged: Minor performance improvement.

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **eae4f6f**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1249 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/QNHNpqVzqP4Y040T511Nj2ddghuJbc3r)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **54ebfcc**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1249 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/21Q7Zv4NHDy0wUGKAidKgAjA9Q9koFYG)

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java [hunk 1]
```diff
@@ -414,8 +414,7 @@ public synchronized long obtainSnapshotTimestamp() {
             // Otherwise, fetch a read token from the sequencer the linearize
             // ourselves against.
             long currentTail = builder.runtime
-                    .getSequencerView().nextToken(Collections.emptySet(),
-                            0).getToken().getTokenValue();
+                    .getSequencerView().cachedTail().getToken().getTokenValue();
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: False. Test it with a debugger, dump the stream cache map.

File: runtime/src/main/java/org/corfudb/runtime/view/CachedSequencerView.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        return cacheToken(null, super.conditional(info, streams));
+    }
+
+    @Override
+    public TokenResponse next(UUID... streams) {
+        return cacheToken(null, super.next(streams));
+    }
+
+    @Override
+    public TokenResponse tail(UUID... streams) {
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/CachedSequencerView.java]: Linearizable reads become cached reads.

You can run this test after you fix the bug I mentioned a couple of times already:

@Test
    public void largeStreamWrite() throws Exception {

        CorfuRuntime rt = new CorfuRuntime("localhost:9000").connect();
        rt.getParameters().setSequencerCacheEnabled(true);

        CorfuRuntime rt2 = new CorfuRuntime("localhost:9000").connect();

        Map<String, String> map1 = rt.getObjectsView().build().setStreamName("s1").setType(SMRMap.class).open();
        Map<String, String> map2 = rt2.getObjectsView().build().setStreamName("s1").setType(SMRMap.class).open();

        map1.put("k0", "v0");
        map2.put("k1", "v1");
        assertThat(map1.get("k1")).isEqualTo("v1");
    }

File: runtime/src/main/java/org/corfudb/runtime/view/CachedSequencerView.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            return tail(streams);
+        }
+
+        // If the request is for a stream tail
+        if (streams.length == 1) {
+            // Get the accumulator for the stream
+            LongAccumulator streamTail = streamTailCache.get(streams[0]);
+            // We don't have this stream cached, go to the sequencer.
+            if (streamTail == null) {
+                return tail(streams);
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/CachedSequencerView.java]: Yea, I did. I think you need to reread your code.

LongAccumulator streamTailAccumulator = streamTailCache.getOrDefault(k, new LongAccumulator(Long::max, tail));

When the streamTailCache is empty, you create a new accumulator but never insert it in the map. In other words, the cache is never populated. Therefore, this caching mechanism doesn't work, or it works with 0% hit-rate.

**Comment by no2chem** [comment]
@Maithem I have to apologize, I guess the version that was pushed is way different than the version on my machine, it's been over a month, sorry. I pushed whatever the current state of that branch is from my machine...

I believe that the current version has a benchmark that shows the performance difference.

**Pull request closed without merging**